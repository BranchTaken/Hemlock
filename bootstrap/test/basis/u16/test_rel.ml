open! Basis.Rudiments
open! Basis
open U16
open Format

let test () =
  let fn x y = begin
    printf "cmp %a %a -> %a\n" pp_x x pp_x y Cmp.pp (cmp x y);
    printf "%a >= %a -> %b\n" pp_x x pp_x y (x >= y);
    printf "%a <= %a -> %b\n" pp_x x pp_x y (x <= y);
    printf "%a = %a -> %b\n" pp_x x pp_x y (x = y);
    printf "%a > %a -> %b\n" pp_x x pp_x y (x > y);
    printf "%a < %a -> %b\n" pp_x x pp_x y (x < y);
    printf "%a <> %a -> %b\n" pp_x x pp_x y (x <> y);
    printf "ascending %a %a -> %a\n" pp_x x pp_x y Cmp.pp (ascending x y);
    printf "descending %a %a -> %a\n" pp_x x pp_x y Cmp.pp (descending x y);
  end in
  fn (kv 0) (kv 0x8000);
  printf "\n";
  fn (kv 0) (kv 0xffff);
  printf "\n";
  fn (kv 0x8000) (kv 0xffff);
  let fn2 t min max = begin
    printf "\n";
    printf "clamp ~min:%a ~max:%a %a -> %a\n"
      pp_x min pp_x max pp_x t pp_x (clamp ~min ~max t);
    printf "between ~low:%a ~high:%a %a -> %b\n"
      pp_x min pp_x max pp_x t (between ~low:min ~high:max t);
  end in
  fn2 (kv 0x7ffe) (kv 0x7fff) (kv 0x8001);
  fn2 (kv 0x7fff) (kv 0x7fff) (kv 0x8001);
  fn2 (kv 0x8000) (kv 0x7fff) (kv 0x8001);
  fn2 (kv 0x8001) (kv 0x7fff) (kv 0x8001);
  fn2 (kv 0x8002) (kv 0x7fff) (kv 0x8001)

let _ = test ()
