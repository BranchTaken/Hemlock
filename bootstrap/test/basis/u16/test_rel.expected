cmp 0x0000u16 0x8000u16 -> Lt
0x0000u16 >= 0x8000u16 -> false
0x0000u16 <= 0x8000u16 -> true
0x0000u16 = 0x8000u16 -> false
0x0000u16 > 0x8000u16 -> false
0x0000u16 < 0x8000u16 -> true
0x0000u16 <> 0x8000u16 -> true
ascending 0x0000u16 0x8000u16 -> Lt
descending 0x0000u16 0x8000u16 -> Gt

cmp 0x0000u16 0xffffu16 -> Lt
0x0000u16 >= 0xffffu16 -> false
0x0000u16 <= 0xffffu16 -> true
0x0000u16 = 0xffffu16 -> false
0x0000u16 > 0xffffu16 -> false
0x0000u16 < 0xffffu16 -> true
0x0000u16 <> 0xffffu16 -> true
ascending 0x0000u16 0xffffu16 -> Lt
descending 0x0000u16 0xffffu16 -> Gt

cmp 0x8000u16 0xffffu16 -> Lt
0x8000u16 >= 0xffffu16 -> false
0x8000u16 <= 0xffffu16 -> true
0x8000u16 = 0xffffu16 -> false
0x8000u16 > 0xffffu16 -> false
0x8000u16 < 0xffffu16 -> true
0x8000u16 <> 0xffffu16 -> true
ascending 0x8000u16 0xffffu16 -> Lt
descending 0x8000u16 0xffffu16 -> Gt

clamp ~min:0x7fffu16 ~max:0x8001u16 0x7ffeu16 -> 0x7fffu16
between ~low:0x7fffu16 ~high:0x8001u16 0x7ffeu16 -> false

clamp ~min:0x7fffu16 ~max:0x8001u16 0x7fffu16 -> 0x7fffu16
between ~low:0x7fffu16 ~high:0x8001u16 0x7fffu16 -> true

clamp ~min:0x7fffu16 ~max:0x8001u16 0x8000u16 -> 0x8000u16
between ~low:0x7fffu16 ~high:0x8001u16 0x8000u16 -> true

clamp ~min:0x7fffu16 ~max:0x8001u16 0x8001u16 -> 0x8001u16
between ~low:0x7fffu16 ~high:0x8001u16 0x8001u16 -> true

clamp ~min:0x7fffu16 ~max:0x8001u16 0x8002u16 -> 0x8001u16
between ~low:0x7fffu16 ~high:0x8001u16 0x8002u16 -> false
